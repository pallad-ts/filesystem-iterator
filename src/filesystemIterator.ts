import * as fs from 'fs';
import * as path from 'path';

export function filesystemIterator(dir: string, options?: filesystemIterator.Options.FromUser): AsyncGenerator<string> {
    const opts: filesystemIterator.Options = {
        filter: options?.filter,
        skipDirectories: options?.skipDirectories ?? false,
        maxDepth: options?.maxDepth ?? Infinity
    };

    const iterator = async function* (dir: string, depth: number): AsyncGenerator<string> {
        const handle = await fs.promises.opendir(dir);
        while (true) {
            const entry = await handle.read();
            if (entry === null) {
                return;
            }

            const currentPath = path.join(dir, entry.name);
            if (entry.isDirectory()) {
                if (depth + 1 <= opts.maxDepth) {
                    yield* iterator(currentPath, depth + 1);
                }
                if (opts.skipDirectories) {
                    continue;
                }
            }

            if (opts.filter && !opts.filter(currentPath)) {
                continue;
            }
            yield currentPath;
        }
    };

    return iterator(dir, 0);
}

export namespace filesystemIterator {
    export interface Options {
        maxDepth: number;
        skipDirectories: boolean;
        filter?: (path: string) => boolean
    }

    export namespace Options {
        export type FromUser = Partial<Pick<Options, 'maxDepth' | 'skipDirectories'>> & Pick<Options, 'filter'>;
    }
}