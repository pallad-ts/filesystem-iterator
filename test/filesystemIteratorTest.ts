import {filesystemIterator} from "@src/filesystemIterator";
import * as path from 'path';


describe('filesystemIterator', () => {

    const ROOT_DIR = path.join(__dirname, './fixtures/example');

    async function iteratorToArray(gen: AsyncGenerator<string>) {
        const results = [];
        for await (const entry of gen) {
            results.push(path.relative(ROOT_DIR, entry));
        }
        return results;
    }

    it('all files', async () => {
        const files = await iteratorToArray(filesystemIterator(ROOT_DIR));

        expect(files)
            .toMatchSnapshot();
    });

    it('skipping directories', async () => {
        const files = await iteratorToArray(filesystemIterator(ROOT_DIR, {skipDirectories: true}));
        expect(files)
            .toMatchSnapshot();
    });

    it('max depth', async () => {
        const files = await iteratorToArray(filesystemIterator(ROOT_DIR, {maxDepth: 1}));
        expect(files)
            .toMatchSnapshot();
    });

    it('using filter', async () => {
        const files = await iteratorToArray(filesystemIterator(ROOT_DIR, {
            filter(x) {
                return x.includes('more');
            }
        }));
        expect(files)
            .toMatchSnapshot();
    });

    it('using filter with skipping directories', async () => {
        const files = await iteratorToArray(filesystemIterator(ROOT_DIR, {
            filter(x) {
                return x.includes('more');
            },
            skipDirectories: true
        }));
        expect(files)
            .toMatchSnapshot();
    })
});