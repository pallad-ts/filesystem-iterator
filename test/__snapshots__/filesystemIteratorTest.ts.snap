// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filesystemIterator all files 1`] = `
Array [
  "file.txt",
  "nested/file2.txt",
  "nested/even/file3.txt",
  "nested/even/more/file4.txt",
  "nested/even/more",
  "nested/even",
  "nested",
]
`;

exports[`filesystemIterator max depth 1`] = `
Array [
  "file.txt",
  "nested/file2.txt",
  "nested/even",
  "nested",
]
`;

exports[`filesystemIterator skipping directories 1`] = `
Array [
  "file.txt",
  "nested/file2.txt",
  "nested/even/file3.txt",
  "nested/even/more/file4.txt",
]
`;

exports[`filesystemIterator using filter 1`] = `
Array [
  "nested/even/more/file4.txt",
  "nested/even/more",
]
`;

exports[`filesystemIterator using filter with skipping directories 1`] = `
Array [
  "nested/even/more/file4.txt",
]
`;
